taichi/backends/opengl/codegen_opengl.cpp

```diff
@@ -291,6 +301,77 @@ class KernelGen : public IRVisitor {
          parent->short_name(),
          struct_compiled_->snode_map.at(parent_type).elem_stride,
          stmt->input_index->short_name(), stmt->snode->node_type_name);
+    if (stmt->activate) {
+      if (stmt->snode->type == SNodeType::dense) {
+        // do nothing
+      } else if (stmt->snode->type == SNodeType::dynamic) {
+        emit("atomicMax(_data_i32_[{} >> 2], {} + 1); // dynamic activate",
+             get_snode_meta_address(stmt->snode),
+             stmt->input_index->short_name());
+      } else {
+        TI_NOT_IMPLEMENTED
+      }
+    }
+  }
+
+  void visit(SNodeOpStmt *stmt) override {  // IAPR?
+    if (stmt->op_type == SNodeOpType::activate) {
```

Comment:
FYI, I don't think `SNodeOpType::activate` will be directly used. E.g. https://github.com/taichi-dev/taichi/blob/37976e750e1579ff9328f3d25f1a63cb3b3292e2/taichi/codegen/codegen_llvm.cpp#L904-L925 There isn't a handling for `::activate`.

IIUC, `::activate` is always translated into an `SNodeLookupStmt` to decide whether this is a strong or weak lookup. But I feel like it's fine to leave it here. Maybe we will have something like `ti.activate(x, [i, j, k])` one day? @yuanming-hu

Reply:
Great, feel free to open an issue in `ti.activate` and assign me later.

Reply:
Thanks for the discussions. I think having `ti.activate` is a good idea.